File: app/actors/StocksActor.java
package actors;

import akka.actor.AbstractActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import stocks.Stock;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collector;
import java.util.stream.Collectors;

/**
 * This actor contains a set of stocks internally that may be used by
 * all websocket clients.
 */
public class StocksActor extends AbstractActor {

    private final Map<String, Stock> stocksMap = new HashMap<>();

    private final LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(Messages.WatchStocks.class, watchStocks -> {
                    Set<Stock> stocks = watchStocks.symbols.stream()
                            .map(symbol -> stocksMap.compute(symbol, (k, v) -> new Stock(k)))
                            .collect(Collectors.toSet());
                    sender().tell(new Messages.Stocks(stocks), self());
                }).build();
    }
}

File: app/actors/StocksActor.scala
package actors

import akka.actor.{Actor, ActorLogging}
import akka.event.LoggingReceive
import stocks._

import scala.collection.mutable

/**
 * This actor contains a set of stocks internally that may be used by
 * all websocket clients.
 */
class StocksActor extends Actor with ActorLogging {

  // May want to remove stocks that aren't viewed by any clients...
  private val stocksMap: mutable.Map[StockSymbol, Stock] = mutable.HashMap()

  def receive = LoggingReceive {
    case WatchStocks(symbols) =>
      val stocks = symbols.map(symbol => stocksMap.getOrElseUpdate(symbol, new Stock(symbol)))
      sender ! Stocks(stocks)
  }
}

case class Stocks(stocks: Set[Stock]) {
  require(stocks.nonEmpty, "Must specify at least one stock!")
}

case class WatchStocks(symbols: Set[StockSymbol]) {
  require(symbols.nonEmpty, "Must specify at least one symbol!")
}

